cns/ipampoolmonitor/ipampoolmonitor.go

```diff
@@ -2,130 +2,197 @@
+	MaximumFreeIps int64
 
-	sync.RWMutex
+	mu sync.RWMutex
 }
 
 func NewCNSIPAMPoolMonitor(cnsService cns.HTTPService, requestController requestcontroller.RequestController) *CNSIPAMPoolMonitor {
 	return &CNSIPAMPoolMonitor{
-		initialized: false,
-		cns:         cnsService,
-		rc:          requestController,
+		pendingRelease: false,
+		cns:            cnsService,
+		rc:             requestController,
+	}
+}
+
+func stopReconcile(ch <-chan struct{}) bool {
+	select {
+	case <-ch:
+		return true
+	default:
 	}
+
+	return false
 }
 
-// TODO: add looping and cancellation to this, and add to CNS MAIN
-func (pm *CNSIPAMPoolMonitor) Start() error {
-
-	if pm.initialized {
-		availableIPConfigs := pm.cns.GetAvailableIPConfigs()
-		rebatchAction := pm.checkForResize(len(availableIPConfigs))
-		switch rebatchAction {
-		case increasePoolSize:
-			return pm.increasePoolSize()
-		case decreasePoolSize:
-			return pm.decreasePoolSize()
+func (pm *CNSIPAMPoolMonitor) Start(ctx context.Context, poolMonitorRefreshMilliseconds int) error {
+	logger.Printf("[ipam-pool-monitor] Starting CNS IPAM Pool Monitor")
+
+	ticker := time.NewTicker(time.Duration(poolMonitorRefreshMilliseconds) * time.Millisecond)
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("CNS IPAM Pool Monitor received cancellation signal")
+	case <-ticker.C:
+		err := pm.Reconcile()
+		if err != nil {
+			logger.Printf("[ipam-pool-monitor] Reconcile failed with err %v", err)
 		}
 	}
 
 	return nil
 }
 
-// UpdatePoolLimitsTransacted called by request controller on reconcile to set the batch size limits
-func (pm *CNSIPAMPoolMonitor) UpdatePoolLimitsTransacted(scalarUnits cns.ScalarUnits) {
-	pm.Lock()
-	defer pm.Unlock()
-	pm.scalarUnits = scalarUnits
+func (pm *CNSIPAMPoolMonitor) Reconcile() error {
+	cnsPodIPConfigCount := len(pm.cns.GetPodIPConfigState())
+	allocatedPodIPCount := len(pm.cns.GetAllocatedIPConfigs())
+	pendingReleaseIPCount := len(pm.cns.GetPendingReleaseIPConfigs())
+	availableIPConfigCount := len(pm.cns.GetAvailableIPConfigs()) // TODO: add pending allocation count to real cns
+	freeIPConfigCount := int64(availableIPConfigCount + (int(pm.cachedSpec.RequestedIPCount) - cnsPodIPConfigCount))
 
-	// TODO rounding?
-	pm.MinimumFreeIps = int(pm.scalarUnits.BatchSize * (pm.scalarUnits.RequestThresholdPercent / 100))
-	pm.MaximumFreeIps = int(pm.scalarUnits.BatchSize * (pm.scalarUnits.ReleaseThresholdPercent / 100))
+	logger.Printf("[ipam-pool-monitor] Checking pool for resize, Pool Size: %v, Goal Size: %v, BatchSize: %v, MinFree: %v, MaxFree:%v, Allocated: %v, Available: %v, Pending Release: %v, Free: %v", cnsPodIPConfigCount, pm.cachedSpec.RequestedIPCount, pm.scalarUnits.BatchSize, pm.MinimumFreeIps, pm.MaximumFreeIps, allocatedPodIPCount, availableIPConfigCount, pendingReleaseIPCount, freeIPConfigCount)
 
-	pm.initialized = true
-}
+	// if there's a pending change to the spec count, and the pending release state is nonzero,
+	// skip so we don't thrash the UpdateCRD
```

Comment:
It is okay to updateCRD and request for more from CNS. We wont be able to update the request anyways allocatedCount wont increase until the next batch is ready. We dont need this check